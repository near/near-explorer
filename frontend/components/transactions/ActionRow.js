import Link from "next/link";

import { Row, Col } from "react-bootstrap";

import { DataConsumer } from "../utils/DataProvider";
import Timer from "../utils/Timer";
import actionIcons from "./ActionIcons";

import ActionMessage from "./ActionMessage";

const ActionRow = props => {
  const { transaction, action } = props;
  let actionKind, actionArgs;
  if (typeof action === "string") {
    actionKind = action;
    actionArgs = {};
  } else {
    actionKind = Object.keys(action)[0];
    actionArgs = action[actionKind];
  }
  const ActionIcon = actionIcons[actionKind];
  const viewMode = props.viewMode || "sparse";
  return (
    <Row
      noGutters="true"
      className={`action-${viewMode}-row mx-0 ${
        props.className === undefined ? "" : props.className
      }`}
    >
      {viewMode === "compact" ? (
        <Col xs="1" className="actions-icon-col">
          <div className="action-row-img">{ActionIcon && <ActionIcon />}</div>
        </Col>
      ) : (
        <>
          <Col md="auto" xs="1" className="pr-0">
            {ActionIcon}
          </Col>
          <Col md="auto" xs="1" className="pr-0">
            <img
              src="/static/images/icon-arrow-right.svg"
              style={{ width: "10px" }}
            />
          </Col>
        </>
      )}
      <Col md="11" xs="11" className="action-row-details">
        <Row noGutters="true">
          <Col md="8" xs="7">
            <Row noGutters="true">
              <Col className="action-row-title">
                <ActionMessage
                  transaction={transaction}
                  actionKind={actionKind}
                  actionArgs={actionArgs}
                />
              </Col>
            </Row>
            <Row noGutters="true">
              <Col className="action-row-text">by @{transaction.signerId}</Col>
            </Row>
          </Col>
          <Col md="4" xs="5" className="ml-auto text-right">
            <Row>
              <Col className="action-row-txid">
                {transaction.hash !== undefined
                  ? `${transaction.hash.substring(0, 7)}...`
                  : null}
              </Col>
            </Row>
            <Row>
              <Col className="action-row-timer">
                <span className="action-row-timer-status">
                  {transaction.status}
                </span>
                &nbsp;&nbsp;
                <Timer time={transaction.blockTimestamp} />
              </Col>
            </Row>
          </Col>
        </Row>
      </Col>
      <style jsx global>{`
        .action-sparse-row {
          padding-top: 10px;
          padding-bottom: 10px;
          border-top: solid 2px #f8f8f8;
        }

        .action-compact-row .action-row-details {
          border-bottom: 2px solid #f8f8f8;
          margin-bottom: 15px;
          padding-bottom: 8px;
        }

        .action-row-img {
          width: 24px;
          height: 24px;
          border: solid 2px #f8f8f8;
          background-color: #ffffff;
          border-radius: 50%;
          margin-right: 8px;
          text-align: center;
          line-height: 1.1;
        }

        .action-row-bottom {
          border-bottom: solid 2px #f8f8f8;
        }

        .action-row-title {
          font-family: BentonSans;
          font-size: 14px;
          font-weight: 500;
          line-height: 1.29;
          color: #24272a;
        }

        .action-row-text {
          font-family: BentonSans;
          font-size: 12px;
          font-weight: 500;
          line-height: 1.5;
          color: #999999;
        }

        .action-row-txid {
          font-family: BentonSans;
          font-size: 14px;
          font-weight: 500;
          line-height: 1.29;
          color: #0072ce;
        }

        .action-row-timer {
          font-family: BentonSans;
          font-size: 12px;
          color: #999999;
          font-weight: 100;
        }

        .action-row-timer-status {
          font-weight: 500;
        }
      `}</style>
    </Row>
  );
};

export default ActionRow;
