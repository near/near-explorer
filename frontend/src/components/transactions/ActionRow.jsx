import { Row, Col } from "react-bootstrap";

import Timer from "../utils/Timer";
import actionIcons from "./ActionIcons";
import ActionMessage from "./ActionMessage";

export default props => {
  const { transaction, action } = props;
  let actionKind, actionArgs;
  if (typeof action === "string") {
    actionKind = action;
    actionArgs = {};
  } else {
    actionKind = Object.keys(action)[0];
    actionArgs = action[actionKind];
  }
  const ActionIcon = actionIcons[actionKind];
  const viewMode = props.viewMode || "sparse";
  return (
    <Row
      noGutters="true"
      className={`action-${viewMode}-row mx-0 ${
        props.className === undefined ? "" : props.className
      }`}
    >
      <Col xs="1" className="actions-icon-col">
        <div className="action-row-img">{ActionIcon && <ActionIcon />}</div>
        {viewMode === "sparse" ? (
          <img
            src="/static/images/icon-arrow-right.svg"
            className="action-row-toggler"
          />
        ) : null}
      </Col>
      <Col md="11" xs="11" className="action-row-details">
        <Row noGutters="true">
          <Col md="8" xs="7">
            <Row noGutters="true">
              <Col className="action-row-title">
                <ActionMessage
                  transaction={transaction}
                  actionKind={actionKind}
                  actionArgs={actionArgs}
                />
              </Col>
            </Row>
            <Row noGutters="true">
              <Col className="action-row-text">by @{transaction.signerId}</Col>
            </Row>
          </Col>
          <Col md="4" xs="5" className="ml-auto text-right">
            <Row>
              <Col className="action-row-txid">
                {transaction.hash !== undefined
                  ? `${transaction.hash.substring(0, 7)}...`
                  : null}
              </Col>
            </Row>
            <Row>
              <Col className="action-row-timer">
                <span className="action-row-timer-status">{`Completed`}</span>{" "}
                <Timer time={transaction.blockTimestamp} />
              </Col>
            </Row>
          </Col>
        </Row>
      </Col>
      <style jsx global>{`
        .action-sparse-row {
          padding-top: 10px;
          padding-bottom: 10px;
          border-top: solid 2px #f8f8f8;
        }

        .action-compact-row .action-row-details {
          border-bottom: 2px solid #f8f8f8;
          margin-bottom: 15px;
          padding-bottom: 8px;
        }

        .action-compact-row .action-row-img {
          width: 24px;
          height: 24px;
          border: solid 2px #f8f8f8;
          background-color: #ffffff;
          border-radius: 50%;
          margin-right: 8px;
          text-align: center;
          line-height: 1.1;
        }

        .action-sparse-row .action-row-img {
          margin: 10px 10px 10px 20px;
          display: inline;
        }

        .action-sparse-row .action-row-img svg {
          height: 20px;
          width: 20px;
        }

        .action-compact-row .action-row-img svg {
          height: 12px;
          width: 12px;
        }

        .action-sparse-row .action-row-toggler {
          width: 10px;
          margin: 10px;
        }

        .action-row-bottom {
          border-bottom: solid 2px #f8f8f8;
        }

        .action-row-title {
          font-family: BentonSans;
          font-size: 14px;
          font-weight: 500;
          line-height: 1.29;
          color: #24272a;
        }

        .action-row-text {
          font-family: BentonSans;
          font-size: 12px;
          font-weight: 500;
          line-height: 1.5;
          color: #999999;
        }

        .action-row-txid {
          font-family: BentonSans;
          font-size: 14px;
          font-weight: 500;
          line-height: 1.29;
          color: #0072ce;
        }

        .action-row-timer {
          font-family: BentonSans;
          font-size: 12px;
          color: #999999;
          font-weight: 100;
        }

        .action-row-timer-status {
          font-weight: 500;
        }
      `}</style>
    </Row>
  );
};
