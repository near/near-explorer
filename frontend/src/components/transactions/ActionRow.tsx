import React from "react";

import actionIcons from "./ActionIcons";
import ActionMessage from "./ActionMessage";
import ActionRowBlock, { ViewMode, DetalizationMode } from "./ActionRowBlock";
import * as T from "../../libraries/explorer-wamp/transactions";

export interface Props {
  action: T.Action;
  transaction: T.Transaction;
  viewMode: ViewMode;
  detalizationMode: DetalizationMode;
  className: string;
  showDetails?: boolean;
}

export default class extends React.Component<Props> {
  static defaultProps = {
    viewMode: "sparse",
    detalizationMode: "detailed",
    className: "",
    showDetails: false
  };

  render() {
    const {
      viewMode,
      detalizationMode,
      className,
      transaction,
      action,
      showDetails
    } = this.props;
    const ActionIcon = actionIcons[action.kind];
    return (
      <ActionRowBlock
        viewMode={viewMode}
        detalizationMode={detalizationMode}
        className={className}
        transaction={transaction}
        icon={ActionIcon && <ActionIcon />}
        title={
          <ActionMessage
            transaction={transaction}
            actionKind={action.kind}
            actionArgs={action.args}
            showDetails={showDetails}
          />
        }
      />
    );
  }
}
