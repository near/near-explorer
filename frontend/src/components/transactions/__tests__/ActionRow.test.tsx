import renderer from "react-test-renderer";

import ActionRow from "../ActionRow";

const TRANSACTION = {
  hash: "BvJeW6gnFjkCBKCsRNEBrRLDQCFZNxLAi6uXzmLaVrrj",
  signerId: "signer.test",
  receiverId: "receiver.test",
  blockTimestamp: +new Date(2019, 1, 1)
};

describe("<ActionRow />", () => {
  it("renders sparsely by default", () => {
    expect(
      renderer.create(
        <ActionRow transaction={TRANSACTION} action={"CreateAccount"} />
      )
    ).toMatchSnapshot();
  });

  it("renders compact", () => {
    expect(
      renderer.create(
        <ActionRow
          viewMode="compact"
          transaction={TRANSACTION}
          action={"CreateAccount"}
        />
      )
    ).toMatchSnapshot();
  });
});
