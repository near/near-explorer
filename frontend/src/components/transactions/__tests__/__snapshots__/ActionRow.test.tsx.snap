// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ActionRow /> renders compact 1`] = `
<div
  className="action-compact-row mx-0  row no-gutters"
>
  <div
    className="actions-icon-col col-1"
  >
    <div
      className="jsx-2818130875 action-row-img"
    >
      <svg
        className="jsx-2818130875"
        viewBox="0 0 18 18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.26 9l-.34-.45a4.5 4.5 0 1 0-3.84 0L6.74 9A7 7 0 0 0 0 16v2h18v-2a7 7 0 0 0-6.74-7z"
          fill="#0071ce"
        />
      </svg>
    </div>
  </div>
  <div
    className="action-row-details col-md-11 col-11"
  >
    <div
      className="row no-gutters"
    >
      <div
        className="col-md-8 col-7"
      >
        <div
          className="row no-gutters"
        >
          <div
            className="action-row-title col"
          >
            New account created: @receiver.test
          </div>
        </div>
        <div
          className="row no-gutters"
        >
          <div
            className="action-row-text col"
          >
            by @
            signer.test
          </div>
        </div>
      </div>
      <div
        className="ml-auto text-right col-md-4 col-5"
      >
        <div
          className="row"
        >
          <div
            className="action-row-txid col"
          >
            BvJeW6g...
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="action-row-timer col"
          >
            <span
              className="jsx-2818130875 action-row-timer-status"
            >
              Completed
            </span>
             
            <span>
              1s ago
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`<ActionRow /> renders sparsely by default 1`] = `
<div
  className="action-sparse-row mx-0  row no-gutters"
>
  <div
    className="actions-icon-col col-1"
  >
    <div
      className="jsx-2818130875 action-row-img"
    >
      <svg
        className="jsx-2818130875"
        viewBox="0 0 18 18"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M11.26 9l-.34-.45a4.5 4.5 0 1 0-3.84 0L6.74 9A7 7 0 0 0 0 16v2h18v-2a7 7 0 0 0-6.74-7z"
          fill="#0071ce"
        />
      </svg>
    </div>
    <img
      className="jsx-2818130875 action-row-toggler"
      src="/static/images/icon-arrow-right.svg"
    />
  </div>
  <div
    className="action-row-details col-md-11 col-11"
  >
    <div
      className="row no-gutters"
    >
      <div
        className="col-md-8 col-7"
      >
        <div
          className="row no-gutters"
        >
          <div
            className="action-row-title col"
          >
            New account created: @receiver.test
          </div>
        </div>
        <div
          className="row no-gutters"
        >
          <div
            className="action-row-text col"
          >
            by @
            signer.test
          </div>
        </div>
      </div>
      <div
        className="ml-auto text-right col-md-4 col-5"
      >
        <div
          className="row"
        >
          <div
            className="action-row-txid col"
          >
            BvJeW6g...
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="action-row-timer col"
          >
            <span
              className="jsx-2818130875 action-row-timer-status"
            >
              Completed
            </span>
             
            <span>
              1s ago
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
